tokio-util/src/codec/framed_impl.rs

```diff
@@ -120,42 +120,101 @@ where
-            // readable, it can be assumed that the decoder will never become
-            // readable again, at which point the stream is terminated.
+            // Repeatedly call `decode` or `decode_eof` as long as long as the buffer
+            // is "readable". Readable means that it _might_ contain data that could be consumed
+            // by `decode` or `decode_eof`. Both `decode` and `decode_eof` signal that the buffer
+            // is no longer readable for them by returning `None`.
+            // If `decode` signals that it couldn't read from the buffer and the upstream has
+            // returned eof, `decode_eof` will attemp to decode the remainder as closing frames.
+            // The buffer might become readable again if the underlying AsyncRead is a FIFO or file.
+            // However, we still want to make sure that upon encountering an eof
+            // we actually finish emmiting all it's associated `decode_eof` frames.
+            // Furthermore, we don't want to emit any `decode_eof` frames on retried
+            // reads after an eof unless we've actually read more data.
             if state.is_readable {
+                // pausing or framing
                 if state.eof {
+                    // pausing
                     let frame = pinned.codec.decode_eof(&mut state.buffer)?;
+                    if frame.is_none() {
+                        state.is_readable = false; // prepare pausing -> paused
+                    }
+                    // implicit pausing -> pausing or pausing -> paused
                     return Poll::Ready(frame.map(Ok));
                 }
 
+                // framing
                 trace!("attempting to decode a frame");
 
                 if let Some(frame) = pinned.codec.decode(&mut state.buffer)? {
                     trace!("frame decoded from buffer");
+                    // implicit framing -> framing
                     return Poll::Ready(Some(Ok(frame)));
                 }
 
+                // framing -> reading
                 state.is_readable = false;
             }
-
-            assert!(!state.eof);
-
-            // Otherwise, try to read more data and try again. Make sure we've
-            // got room for at least one byte to read to ensure that we don't
-            // get a spurious 0 that looks like EOF
-            state.buffer.reserve(1);
+            // reading or stopped
+            // If we can't build a frame yet, try to read more data and try again.
+            // Make sure we've got room for at least one byte to read to ensure
+            // that we don't get a spurious 0 that looks like eof.
+            // Only allocate if we actually need it to prevent slowly leaking bytes,
+            // in less fortunate code that repeatedly polls an eof file.
+            if state.buffer.capacity() == state.buffer.len() {
+                state.buffer.reserve(1);
+            }
```

Comment:
Reserve doesn't do anything if there is enough capacity, so this `if` is not necessary.

Reply:
Thanks for the correction! I misread the docs and mistook the `additional` for an emphasis instead of simple arg reference. 😅 🙇 
I'll gladly rip it out again! 😁

Reply:
Removed :D!